pulsar-log4j2-appender/src/test/java/org/apache/pulsar/log4j2/appender/AbstractPulsarAppenderTest.java

```diff
@@ -0,0 +1,167 @@
+        doReturn(clientBuilder).when(clientBuilder).serviceUrl(anyString());
+        doReturn(clientBuilder).when(clientBuilder).loadConf(any());
+        doReturn(clientBuilder).when(clientBuilder).authentication(any());
+        doReturn(clientBuilder).when(clientBuilder).tlsTrustStorePassword(anyString());
+        doReturn(clientBuilder).when(clientBuilder).tlsTrustStorePath(anyString());
+        doReturn(clientBuilder).when(clientBuilder).tlsTrustCertsFilePath(anyString());
+        doReturn(clientBuilder).when(clientBuilder).tlsTrustStoreType(anyString());
+        doReturn(clientBuilder).when(clientBuilder).useKeyStoreTls(anyBoolean());
+        doReturn(clientBuilder).when(clientBuilder).enableTlsHostnameVerification(anyBoolean());
+        doReturn(clientBuilder).when(clientBuilder).allowTlsInsecureConnection(anyBoolean());
+
+        producerBuilder = mock(ProducerBuilder.class);
+        when(client.newProducer()).thenReturn(producerBuilder);
+        doReturn(producerBuilder).when(producerBuilder).topic(anyString());
+        doReturn(producerBuilder).when(producerBuilder).producerName(anyString());
+        doReturn(producerBuilder).when(producerBuilder).enableBatching(anyBoolean());
+        doReturn(producerBuilder).when(producerBuilder).batchingMaxPublishDelay(anyLong(), any(TimeUnit.class));
+        doReturn(producerBuilder).when(producerBuilder).blockIfQueueFull(anyBoolean());
+        doReturn(producerBuilder).when(producerBuilder).loadConf(any());
+        doReturn(producer).when(producerBuilder).create();
+        doReturn(CompletableFuture.completedFuture(null)).when(producer).closeAsync();
+
+        when(producer.newMessage()).then(invocation -> new MockedMessageBuilder());
+        when(producer.send(any(byte[].class)))
+                .thenAnswer(invocationOnMock -> {
+                    Message<byte[]> msg = invocationOnMock.getArgument(0);
+                    synchronized (history) {
+                        history.add(msg);
+                    }
+                    return null;
+                });
+
+        when(producer.sendAsync(any(byte[].class)))
+                .thenAnswer(invocationOnMock -> {
+                    Message<byte[]> msg = invocationOnMock.getArgument(0);
+                    synchronized (history) {
+                        history.add(msg);
+                    }
+                    CompletableFuture<MessageId> future = new CompletableFuture<>();
+                    future.complete(mock(MessageId.class));
+                    return future;
+                });
+
+        PulsarManager.PULSAR_CLIENT_BUILDER = () -> clientBuilder;
+
+        ctx = Configurator.initialize(
+                "PulsarAppenderTest",
+                getClass().getClassLoader(),
+                getClass().getClassLoader().getResource(resourceName).toURI());
+    }
+
+    protected LogEvent deserializeLogEvent(final byte[] data) throws IOException, ClassNotFoundException {
```

Comment:
a good follow up of this patch is to add support for encoding log events to JSON.

Java Serialization is unsecure, and also if we support JSON it would be super easy to integrate with the ElasticSearch or the Cassandra Sink

Reply:
JSON serialization works out of the box when Jackson is on the classpath. I have experimented with that in https://github.com/lhotari/pulsar-log4j2-appender-demo#k8s-demo .

Reply:
great